### ISSUE TITLE
Validate impurity and that no storage opcodes are generated in a pure context

### ISSUE THREAD

**Issue by mohammadfawaz** [FuelLabs/sway#1151]

**Comment by otrho** [comment]
I am finally _really_ close to having a PR for this, but I have hit a minor related snag.  We already have checks in the compiler to disallow storage operations in non-contracts (and now the storage declarations too) but that includes libraries. 

If a contract includes a library which does storage then that's OK, but if a script includes that library, no... right?  This is not possible to check in the current framework, we may be compiling a library and we have no idea who is going to use it later.

I want to just allow storage ops in libraries for now, and maybe we can create an issue to add internal purity attributes to libraries as we're compiling them, so that then if they're `use`d in a script we can barf.  Is this fair?

**Comment by otrho** [comment]
OK, either way, the problem now is that if we put storage ops in a library they'll be blocked by an error.
```
error
  --> /usr/home/toby/prj/sway/sway-lib-std/src/storage.sw:8:8
   |
 6 |
 7 | #[storage(write)]
 8 | pub fn store<T>(key: b256, value: T) {
   |        ^^^^^ Impure function inside of non-contract. Contract storage is only accessible from contracts.
 9 |     if !__is_reference_type::<T>() {
10 |         // If copy type, then it's a single word and can be stored with a single SWW.
   |
```
So if a contract, which is allowed storage ops, includes this library it'll still fail.  We need to be able to tell what sort of program is including the library before we throw the failure.

So I'm temporarily enabling storage ops in libraries and suggesting we add a new issue to resolve it properly.

**Comment by mohammadfawaz** [comment]
@adlerjohn you also opened https://github.com/FuelLabs/sway/issues/963 recently. Is there a place where we can look up the desired behavior of each VM instruction depending on its context (i.e. internal v.s. external)? Not every contract instruction has to be forbidden in scripts (like block height as you said).

**Comment by otrho** [comment]
I've mentioned this in the PR https://github.com/FuelLabs/sway/pull/1716#issuecomment-1151616089 but yeah, at the moment, while everything is inlined, it's hard/impossible to know whether a function is in a library or contract or just a local script.  We could tag them as such before inlining, which wouldn't be that hard I imagine.

**Comment by adlerjohn** [comment]
> Is there a place where we can look up the desired behavior of each VM instruction depending on its context (i.e. internal v.s. external)?

Yes, the specs 😂. Jokes aside, I think the only instructions forbidden to scripts are direct state read/write and mint/burn. This potentially warrants further separation of instructions into pure/non-pure/contract methinks.
